File: C#/myLeetCode/Binary_Tree/Traversal mode of Binary tree/iteration/94. Binary Tree Inorder Traversal.cs
// 94. Binary Tree Inorder Traversal
// https://leetcode.com/problems/binary-tree-inorder-traversal/

using System;
using System.Collections.Generic;

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        IList<int> result = new List<int>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode currentNode = root;

        while (currentNode != null || stack.Count > 0) {
            while (currentNode != null) {
                stack.Push(currentNode);
                currentNode = currentNode.left;
            }

            currentNode = stack.Pop();
            result.Add(currentNode.val);
            currentNode = currentNode.right;
        }

        return result;
    }
}

public class Program {
    public static void Main() {
        Solution solution = new Solution();

        TreeNode tree = new TreeNode(1) {
            right = new TreeNode(2) {
                left = new TreeNode(3)
            }
        };

        IList<int> inorderTraversal = solution.InorderTraversal(tree);

        Console.WriteLine("Inorder Traversal:");
        foreach (int val in inorderTraversal) {
            Console.Write(val + " ");
        }
    }
}

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/recursion/94. Binary Tree Inorder Traversal.py
from typing import Optional
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        return self.inorderTraversal(root.left) + [root.val] + self.inorderTraversal(root.right) if root else []

sol = Solution()

# 测试用例
root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
print(sol.inorderTraversal(root))  # 应输出 [1, 3, 2]
